File: lib/utilities/setupLogging.ts
import * as cdk from "aws-cdk-lib";
import * as iam from "aws-cdk-lib/aws-iam";
import * as ecs from "aws-cdk-lib/aws-ecs";
import * as logs from "aws-cdk-lib/aws-logs";

import * as path from "path";

import { Environment } from "../environment";

export default function setupLogging(
  stack: cdk.Stack,
  cluster: ecs.Cluster,
  environment: Environment
) {
  let configuration = {
    retention: logs.RetentionDays.INFINITE,
    removalPolicy: environment.production
      ? cdk.RemovalPolicy.RETAIN
      : cdk.RemovalPolicy.DESTROY,
  };

  let logGroups = {
    error: new logs.LogGroup(stack, "ServerErrorLogGroup", configuration),
    access: new logs.LogGroup(stack, "ServerAccessLogGroup", configuration),
    application: new logs.LogGroup(
      stack,
      "ServerApplicationLogGroup",
      configuration
    ),
  };

  let loggingTaks = new ecs.Ec2TaskDefinition(stack, "LoggingTask", {
    taskRole: new iam.Role(stack, "LoggingRole", {
      assumedBy: new iam.ServicePrincipal("ecs-tasks.amazonaws.com"),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName(
          "CloudWatchAgentServerPolicy"
        ),
      ],
    }),
    volumes: [
      {
        name: "ServerVolume",
        host: { sourcePath: environment.server.storage.deviceMountPoint },
      },
    ],
  });

  let loggingContainer = loggingTaks.addContainer("LoggingContainer", {
    image: ecs.ContainerImage.fromAsset(
      path.join(__dirname, "../../src/server/utilities/logging")
    ),
    memoryLimitMiB: 256,
    environment: {
      LOGS_DIRECTORY_PATH: environment.server.variables.LOGS_DIRECTORY_PATH,
      ERROR_LOG_GROUP_NAME: logGroups.error.logGroupName,
      ACCESS_LOG_GROUP_NAME: logGroups.access.logGroupName,
      APPLICATION_LOG_GROUP_NAME: logGroups.application.logGroupName,
    },
    logging: ecs.LogDriver.awsLogs({ streamPrefix: "LoggingLogs" }),
  });

  loggingContainer.addMountPoints({
    sourceVolume: "ServerVolume",
    containerPath: environment.server.storage.containerMountPoint,
    readOnly: true,
  });

  new ecs.Ec2Service(stack, "LoggingService", {
    cluster,
    daemon: true,
    taskDefinition: loggingTaks,
  });
}

File: nitecap/utilities/setup_logging.py
import aws_cdk as cdk
import aws_cdk.aws_iam as iam
import aws_cdk.aws_ecs as ecs
import aws_cdk.aws_logs as logs

from ..configuration import Configuration


def setup_logging(cluster: ecs.Cluster, configuration: Configuration):
    stack = cluster.stack

    if configuration.production:
        removal_policy = cdk.RemovalPolicy.RETAIN
    else:
        removal_policy = cdk.RemovalPolicy.DESTROY

    logs_configuration = {
        "retention": logs.RetentionDays.INFINITE,
        "removal_policy": removal_policy,
    }

    log_groups = {
        "ERROR": logs.LogGroup(stack, "ServerErrorLogGroup", **logs_configuration),
        "ACCESS": logs.LogGroup(stack, "ServerAccessLogGroup", **logs_configuration),
        "APPLICATION": logs.LogGroup(
            stack, "ServerApplicationLogGroup", **logs_configuration
        ),
    }

    logging_task = ecs.Ec2TaskDefinition(
        stack,
        "LoggingTask",
        task_role=iam.Role(
            stack,
            "LoggingRole",
            assumed_by=iam.ServicePrincipal("ecs-tasks.amazonaws.com"),
            managed_policies=[
                iam.ManagedPolicy.from_aws_managed_policy_name(
                    "CloudWatchAgentServerPolicy"
                )
            ],
        ),
        volumes=[
            ecs.Volume(
                name="ServerVolume",
                host=ecs.Host(
                    source_path=configuration.server.storage.device_mount_point
                ),
            )
        ],
    )

    logging_container = logging_task.add_container(
        "LoggingContainer",
        image=ecs.ContainerImage.from_asset("nitecap/server/utilities/logging"),
        memory_limit_mib=256,
        environment={
            "LOGS_DIRECTORY_PATH": configuration.server.environment_variables.LOGS_DIRECTORY_PATH,
            "ERROR_LOG_GROUP_NAME": log_groups["ERROR"].log_group_name,
            "ACCESS_LOG_GROUP_NAME": log_groups["ACCESS"].log_group_name,
            "APPLICATION_LOG_GROUP_NAME": log_groups["APPLICATION"].log_group_name,
        },
        logging=ecs.LogDriver.aws_logs(stream_prefix="LoggingLogs"),
    )

    logging_container.add_mount_points(
        ecs.MountPoint(
            source_volume="ServerVolume",
            container_path=configuration.server.storage.container_mount_point,
            read_only=True,
        )
    )

    ecs.Ec2Service(
        stack,
        "LoggingService",
        cluster=cluster,
        daemon=True,
        task_definition=logging_task,
    )
